[KOGITO-3375] - Make the operator to fetch the metrics endpoint to add Prometheus support

vaibhavjainwiz opened the pull request:
Many thanks for submiting your Pull Request :heart:! 

https://issues.redhat.com/browse/KOGITO-3375

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

vaibhavjainwiz commented:
@ricardozanini 
 Could you please review package structure. I am not clear How should I arrange these files. Code to make rest call is at below location :
https://github.com/vaibhavjainwiz/kogito-cloud-operator/blob/99eb0b61dd073e686c25521c6d15a5bba428f8b5/pkg/infrastructure/services/prometheus.go#L70

ricardozanini commented on pkg/infrastructure/kogito_service.go:
```diff
@@ -103,3 +105,18 @@ func getServiceEndpoints(client *client.Client, namespace string, serviceHTTPRou
+		return "", fmt.Errorf("service with name %s not exist for Kogito service instance in given namespace %s", name, namespace)
+	}
+	port := service.Spec.Ports[0]
+	uri := fmt.Sprintf("%s://%s:%d", port.Name, service.Name, port.TargetPort.IntVal)
```
We should use the Kubernetes svc DNS name instead:
https://github.com/kubernetes/dns/blob/master/docs/specification.md#241---aaaaa-records

```go
// resolves to http://dataindex.mynamespace.svc.cluster.local:8080 for example
uri := fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", service.Name, service.Namespace, port.TargetPort.IntVal)
```

Also `portName` can be anything, let's stick with the standard HTTP protocol since it's an internal call.

That's because we are about to change the operator to be cluster scoped instead, meaning that the operator will be in different namespace of its operands.

vaibhavjainwiz replied:
done

ricardozanini commented on pkg/infrastructure/services/prometheus.go:
```diff
@@ -18,66 +18,151 @@ import (
+	if err != nil {
+		return false, err
+	}
+	if resp, err := client.HTTPCli.Do(req); err != nil {
```
Here, instead of performing the request directly we can wait for the Deployment status of the service, if it's ready to receive connections we do it, otherwise just return without error and try in the next reconciliation loop.

The Deployment `status.availableReplicas` > 0 will tell you if it's ready to receive requests. This way we won't keep making failed requests to the deployed service.

vaibhavjainwiz replied:
done

ricardozanini commented on pkg/client/client.go:
```diff
@@ -59,6 +59,7 @@ type Client struct {
 	Discovery              discovery.DiscoveryInterface
 	DeploymentCli          appsv1.AppsV1Interface
 	KubernetesExtensionCli kubernetes.Interface
+	HTTPCli                restclient2.HTTPClient
```
I believe we should not add this attribute to the `Client` structure since it's just a `http.Client` instance dancing around the functions. It's safer to create it once we need it, plus it's not associated with any k8s client whatsoever.

vaibhavjainwiz replied:
I had removed it from `Client` and directly call `http.Get` when required

ricardozanini commented on pkg/client/restclient/http_client.go:
```diff
@@ -0,0 +1,22 @@
+// HTTPClient interface
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
```
I believe we don't need this. Just interact with the `http` package directly. It's designed to be this way.

vaibhavjainwiz replied:
`HTTPClient` interface helps me to mock http call. If we directly initialize `http.Client` in calling method then we will not be able to mock its call to run testcase. This is a same reason I have to put `HTTPCli` inside `client.go`

ricardozanini replied:
You can use the `httptest` package directly. I understand that the Interface approach is used out there, so that might be ok for now. This will be refactored soon.

Now, about the attribute in the client package I'd rather to remove it from there. That package is meant to be a k8s API client handler only.

vaibhavjainwiz replied:
done

ricardozanini reviewed (changes requested):
Looks good! Just some comments about the internal implementation approach. I had another idea to implement this to reuse the "service is ready to receive calls / svc discovery" part, but it's better to bring this offline in a different PR and showcase to you guys. @Kevin-Mok you will likely to reuse some of this code for [KOGITO-3376](https://issues.redhat.com/browse/KOGITO-3376)

sutaakar commented:
@vaibhavjainwiz Please rebase, will do a review once Zanini's comments are addressed.

ricardozanini commented on pkg/apis/app/v1alpha1/kogitoservices_types.go:
```diff
@@ -188,6 +189,12 @@ type KogitoServiceSpec struct {
+	// Create Service monitor instance to connect with Monitoring service
+	// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors=true
+	// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors.displayName="Monitoring"
+	// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors.x-descriptors="urn:alm:descriptor:com.tectonic.ui:label"
```
This won't work since it's a data structure and not a String. I need to take a look at this actually to see if something changed in the new OLM version.

vaibhavjainwiz replied:
I had removed these extra comments. Just keep first line

ricardozanini commented on pkg/infrastructure/kogito_service.go:
```diff
@@ -103,3 +105,20 @@ func getServiceEndpoints(client *client.Client, namespace string, serviceHTTPRou
+	}
+	port := service.Spec.Ports[0]
+	// resolves to http://dataindex.mynamespace.svc.cluster.local:8080 for example
+	uri := fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", service.Name, service.Namespace, port.TargetPort.IntVal)
```
During the weekend I made some tests around this and we can surpress the `svc.cluster.local` part. People can change this name in the kube configuration to differentiate between zones on hybrid installations, so I believe we need to remove it also.

Just this will work:

```suggestion
	uri := fmt.Sprintf("http://%s.%s:%d", service.Name, service.Namespace, port.Port)
```

Also `TargetPort` it's the port which the svc target in the container, not the exposed one. To use the exposed port we need to use `service.Port` instead. Soon we will use the default HTTP port and adding the port to the service won't be needed any more. See: https://issues.redhat.com/browse/KOGITO-3516

ricardozanini replied:
Depends on: #598 to remove the port part.

vaibhavjainwiz replied:
Updated to suggested change

ricardozanini commented on pkg/infrastructure/kogito_service.go:
```diff
@@ -103,3 +105,20 @@ func getServiceEndpoints(client *client.Client, namespace string, serviceHTTPRou
+	port := service.Spec.Ports[0]
+	// resolves to http://dataindex.mynamespace.svc.cluster.local:8080 for example
+	uri := fmt.Sprintf("http://%s.%s.svc.cluster.local:%d", service.Name, service.Namespace, port.TargetPort.IntVal)
+	//uri := fmt.Sprintf("%s://%s:%d", port.Name, service.Name, port.TargetPort.IntVal)
```
```suggestion

```

vaibhavjainwiz replied:
removed commented code

ricardozanini commented on pkg/infrastructure/services/prometheus.go:
```diff
@@ -18,66 +18,157 @@ import (
+func createServiceMonitor(cli *client.Client, kogitoService v1alpha1.KogitoService, scheme *runtime.Scheme) (*monv1.ServiceMonitor, error) {
+	monitoring := kogitoService.GetSpec().GetMonitoring()
+	endPoint := monv1.Endpoint{}
+	endPoint.TargetPort = &intstr.IntOrString{IntVal: getServiceHTTPPort(kogitoService)}
```
I haven't found the `getServiceHTTPPort` function in this PR, but just make sure we are returning the `service.Port` and not `service.TargetPort`.

sutaakar replied:
IMHO the `getServiceHTTPPort` returns a port where Kogito application is running (a bit unfortunate naming) - https://github.com/kiegroup/kogito-cloud-operator/blob/master/pkg/infrastructure/services/deployment.go#L85. 
Needs to be changed here to use service port.

ricardozanini replied:
Yes, this function is getting the port from the spec, getting from the service directly will be more reliable.

vaibhavjainwiz replied:
@ricardozanini @sutaakar 
Previously I done so because exposed port are always equal to kogitoService port but now after implementation of KOGITO-3516, exposed port always be a default port. So rather then loading Service object and fetch port from it, could I directly use default http?
We could use same approach to create kogitoServiceURI as well.
https://github.com/kiegroup/kogito-cloud-operator/blob/605cb71f5d01b94de56516aaa230ec37232fb846/pkg/infrastructure/kogito_service.go#L110

ricardozanini replied:
+1 @vaibhavjainwiz

vaibhavjainwiz replied:
done

ricardozanini reviewed (approved):
Looks good, one minor thing only. Many thanks!

sutaakar commented on pkg/infrastructure/services/deployer.go:
```diff
@@ -205,6 +205,7 @@ func (s *serviceDeployer) Deploy() (reconcileAfter time.Duration, err error) {
 		}
 		s.generateEventForDeltaResources("Removed", resourceType, delta.Removed)
 	}
+	reconcileAfter, err = configurePrometheus(s.client, s.instance, s.scheme)
```
```suggestion
	// In case required resource needs reconciliation
	if reconcileAfter > 0 {
		return
	}
	reconcileAfter, err = configurePrometheus(s.client, s.instance, s.scheme)
```
This will make sure that Prometheus won't overwrite reconciliation period.

vaibhavjainwiz replied:
I that case we could move this check to `createRequiredResources` itself so that if `createRequiredResources` requires reconcilation then it will return from there and don't make a call `getDeployedResources`

sutaakar commented:
Small BDD adjustment for these changes, can be added as part of this PR - https://github.com/vaibhavjainwiz/kogito-cloud-operator/pull/1

sutaakar commented:
Checked the built operator image on OCP, works as expected.

spolti commented on pkg/infrastructure/services/deployment.go:
```diff
@@ -103,3 +105,12 @@ func setHTTPPortInEnvVar(httpPort int32, kogitoService v1alpha1.KogitoService) {
 	kogitoService.GetSpec().SetEnvs(modifiedEnv)
 }
+
+func isDeploymentAvailable(cli *client.Client, name, namespace string) (bool, error) {
```
will this be called only during reconciliation?

vaibhavjainwiz replied:
yes @spolti . We call this function while configuring prometheus as pre-check. Operator goes to configure prom only when deployment is ready to serve request because operator have to application's `/metrics`  endpoint to check whether application support prom or not

https://github.com/kiegroup/kogito-cloud-operator/blob/640c94390db7cdef2f2e95f258c51013195ebae5/pkg/infrastructure/services/prometheus.go#L42

spolti commented:
looks good, just left a question, and please take a look on the failing test.

radtriste reviewed (approved):
lgtm
Please merge PR adjustment from @sutaakar : https://github.com/vaibhavjainwiz/kogito-cloud-operator/pull/1

vaibhavjainwiz commented:
> lgtm
> Please merge PR adjustment from @sutaakar : [vaibhavjainwiz#1](https://github.com/vaibhavjainwiz/kogito-cloud-operator/pull/1)

done

Kaitou786 commented:
/jenkins test

sutaakar commented:
@vaibhavjainwiz Sorry for wrong information in previous comment....
Right now the ServiceMonitor created by Kogito operator points to wrong port, causing Prometheus to miss Kogito application.
There are two options to specify what port should be used 
- port (https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml#L179) pointing to the service port
- targetPort (https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml#L239) pointing to the pod's port

Right now this implementation use `targetPort`, however it is passing service port there, which is wrong.
So we can either adjust the implementation to use `port` or change the passed port to pod's one.

ricardozanini commented:
> Right now this implementation use `targetPort`, however it is passing service port there, which is wrong.
> So we can either adjust the implementation to use `port` or change the passed port to pod's one.

I'd rather use the service one since we can abstract the exposed port in the container.

Thinking about it now, I think we can get rid of the `spec.httpPort` attribute. There's no practical use. The default http port being served by the service addresses all the issues and make it way simpler to do svc discovery/injection.

@sutaakar @spolti @vaibhavjainwiz wdyt?

spolti commented:
@ricardozanini  if there's no use cas for this property anymore, +1 for removing it.

spolti commented:
@vaibhavjainwiz is the problem reported by @sutaakar addressed?

sutaakar commented:
@ricardozanini fine for me

vaibhavjainwiz commented:
> > Right now this implementation use `targetPort`, however it is passing service port there, which is wrong.
> > So we can either adjust the implementation to use `port` or change the passed port to pod's one.
> 
> I'd rather use the service one since we can abstract the exposed port in the container.
> 
> Thinking about it now, I think we can get rid of the `spec.httpPort` attribute. There's no practical use. The default http port being served by the service addresses all the issues and make it way simpler to do svc discovery/injection.
> 
> @sutaakar @spolti @vaibhavjainwiz wdyt?

Prometheus operator provides both options to create `Service monitor` and `Pod Monitor`. Kogito-operator creates service monitor objects to expose services for scraping so it make sence to provide service pod. 

@ricardozanini I am agree with @sutaakar point to take spec.httpPort as input.

ricardozanini commented:
> @ricardozanini fine for me
> 
> 
> 
> Edit: How do we recognize then what port is used by Kogito application? IMHO the `spec.httpPort` is mainly used for this, to define what port should be set in Deployment as ContainerPort

The question is: why do we need to recognize this? The port should always be defined on 8080 and exposed by the image. This should not change, there's no practical use case for this. 

The operator creates a service exposing this port **through the default HTTP one**. Other services within the cluster will just call the application. 

That's my point. There's no use to change the port, no use case. It's there just because. :)

vaibhavjainwiz commented:
> > @ricardozanini fine for me
> > Edit: How do we recognize then what port is used by Kogito application? IMHO the `spec.httpPort` is mainly used for this, to define what port should be set in Deployment as ContainerPort
> 
> The question is: why do we need to recognize this? The port should always be defined on 8080 and exposed by the image. This should not change, there's no practical use case for this.
> 
> The operator creates a service exposing this port **through the default HTTP one**. Other services within the cluster will just call the application.
> 
> That's my point. There's no use to change the port, no use case. It's there just because. :)

Its spring functionality which allow user to set his desired port through `service.port=9090` prop in `application.properties` file. Atfer setting this property, every time user start its service through `java -jar myapplication.jar`, its embedded tomcat server will start at user provided port. 
Currently I can't think of any specific reason to explicitly change its default port. In my previous project I used to explictly change port values because we have to run multiple java service on our local machine to test some functionality and we can't run containers on dev system due to hardware constraints. 
But in QA/Prod env we always run all services at 8080 port in containers.

ricardozanini commented:
> Its spring functionality which allow user to set his desired port through `service.port=9090` prop in `application.properties` file. Atfer setting this property, every time user start its service through `java -jar myapplication.jar`, its embedded tomcat server will start at user provided port.
> Currently I can't think of any specific reason to explicitly change its default port. In my previous project I used to explictly change port values because we have to run multiple java service on our local machine to test some functionality and we can't run containers on dev system due to hardware constraints.
> But in QA/Prod env we always run all services at 8080 port in containers.

Thanks for sharing your experience, @vaibhavjainwiz. But in this containers scenarios we won't clash ports since there's only one process running within the container and we control it. In case of sidecars injections to share the network from a pod, this other injection pod would require a different port, but again controlled by the operator. There's no use case to change this port any more, only on local environments where you need to share the network resources.

There's no practical usage for this and we are only adding unnecessary code/complexity to our use case. The container will be deployed and served at port 80 by the Kubernetes service. Consuming the Kogito service within the cluster will be straightforward like: `http://name.namespace`.

vaibhavjainwiz commented:
> Thanks for sharing your experience, @vaibhavjainwiz. But in the container scenario we won't clash ports since there's only one process running within the container and we control it. In case of sidecars injections to share the network from a pod, this other injection pod would require a different port, but again controlled by the operator. There's no use case to change this port any more, only on local environments where you need to share the network resources.
> 
> There's no practical usage for this and we are only adding unnecessary code/complexity to our use case. The container will be deployed and served at port 80 by the Kubernetes service. Consuming the Kogito service within the cluster will be straightforward like: `http://name.namespace`.
> 
> And I'd say that we should always force 8080 even if a external configuration like `service.port` is passed since we are forcing the port to be exposed on 8080 in the images:
> 
> ```dockerfile
> EXPOSE 8080
> ```

@ricardozanini Totally agree with your. Inside Image we control service port through env var `HTTP_PORT` but image default exposed port is always 8080.

https://github.com/vaibhavjainwiz/kogito-images/blob/f229b91aeeec09c12a20c3a68010a8f7d40ddde0/modules/kogito-quarkus/added/launch/kogito-quarkus-native.sh#L13-L12

ricardozanini commented:
Thanks for bringing the link, @vaibhavjainwiz :D

Opened: https://issues.redhat.com/browse/KOGITO-3568

vaibhavjainwiz commented:
> @vaibhavjainwiz Sorry for wrong information in previous comment....
> Right now the ServiceMonitor created by Kogito operator points to wrong port, causing Prometheus to miss Kogito application.
> There are two options to specify what port should be used
> 
> * port (https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml#L179) pointing to the service port
> * targetPort (https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml#L239) pointing to the pod's port
> 
> Right now this implementation use `targetPort`, however it is passing service port there, which is wrong.
> So we can either adjust the implementation to use `port` or change the passed port to pod's one.

vaibhavjainwiz commented:
> > > Right now this implementation use `targetPort`, however it is passing service port there, which is wrong.
> > > So we can either adjust the implementation to use `port` or change the passed port to pod's one.
> > 
> > 
> > I'd rather use the service one since we can abstract the exposed port in the container.
> > Thinking about it now, I think we can get rid of the `spec.httpPort` attribute. There's no practical use. The default http port being served by the service addresses all the issues and make it way simpler to do svc discovery/injection.
> > @sutaakar @spolti @vaibhavjainwiz wdyt?
> 
> Prometheus operator provides both options to create `Service monitor` and `Pod Monitor`. Kogito-operator creates service monitor objects to expose services for scraping so it make sence to provide service pod.
> 
> @ricardozanini I am agree with @sutaakar point to take spec.httpPort as input.

@sutaakar Now our expose port is 80 so don't need to define it explicitly inside `ServiceMonitor`. I tested it on my local machine, its generating metrics now.
![image](https://user-images.githubusercontent.com/16335013/95433012-0a127880-096d-11eb-9c21-454b104c489b.png)

ricardozanini commented:
@vaibhavjainwiz this might be a silly question, but can we change the "endpoint" field to reflect the actual port (80)? Or this is the pod IP and port?

vaibhavjainwiz commented:
> @vaibhavjainwiz this might be a silly question, but can we change the "endpoint" field to reflect the actual port (80)? Or this is the pod IP and port?

I had already tried it by explictly setting `EndPoint.Port = 80` but it didn't work. 

its description dictates as :

```
// Name of the service port this endpoint refers to. Mutually exclusive with targetPort.
Port string `json:"port,omitempty"`
```

sutaakar commented:
/jenkins test

sutaakar reviewed (approved):
Looks good to me, lets wait for PR check to finish

ricardozanini commented:
@spolti can be merged
